.circleci/config.yml

```diff
@@ -220,55 +247,135 @@ run_tests_from_wheel: &run_tests_from_wheel
+# Yes, this is horribly redundant. Too lazy to make sense of CircleCI's parameterize 
+# stuff. Might fix when it gets too awful
 jobs:
-  py37_linux:
+  # Pre-commit tests
+  py38_precommit:
+    docker:
+      - image: cimg/python:3.8
+    steps:
+      - checkout
+      - <<: *install_suitesparse
+      - <<: *install_functorch
+      - <<: *build_baspacho_cpu
+      - <<: *run_pre_commit
+
+  py39_precommit:
+    docker:
+      - image: cimg/python:3.9
+    steps:
+      - checkout
+      - <<: *install_suitesparse
+      - <<: *install_functorch
+      - <<: *build_baspacho_cpu
+      - <<: *run_pre_commit
+
+  py310_precommit:
+    docker:
+      - image: cimg/python:3.10
+    steps:
+      - checkout
+      - <<: *install_suitesparse
+      - <<: *install_functorch
+      - <<: *build_baspacho_cpu
+      - <<: *run_pre_commit
+
+  # End-to-end tests
+  py38_e2e_tests:
+    docker:
+      - image: cimg/python:3.8
+    steps:
+      - checkout
+      - <<: *install_suitesparse
+      - <<: *install_functorch
+      - <<: *build_baspacho_cpu
+      - <<: *run_end_to_end_test
+
+  py39_e2e_tests:
     docker:
-      - image: cimg/python:3.7
+      - image: cimg/python:3.9
     steps:
       - checkout
       - <<: *install_suitesparse
       - <<: *install_functorch
       - <<: *build_baspacho_cpu
-      - <<: *install_nox
-      - run:
-          name: "Testing theseus"
-          command: |
-              export NOX_PYTHON_VERSIONS=3.7
-              pip install nox==2020.8.22
-              BASPACHO_ROOT_DIR=~/baspacho nox
-
-  py38_linux:
+      - <<: *run_end_to_end_test
+
+  py310_e2e_tests:
+    docker:
+      - image: cimg/python:3.10
+    steps:
+      - checkout
+      - <<: *install_suitesparse
+      - <<: *install_functorch
+      - <<: *build_baspacho_cpu
+      - <<: *run_end_to_end_test
+
+  # Lie group tests
+  py38_lie_tests:
     docker:
       - image: cimg/python:3.8
     steps:
       - checkout
       - <<: *install_suitesparse
       - <<: *install_functorch
       - <<: *build_baspacho_cpu
-      - <<: *install_nox
-      - run:
-          name: "Testing theseus"
-          command: |
-              export NOX_PYTHON_VERSIONS=3.8
-              pip install nox==2020.8.22
-              BASPACHO_ROOT_DIR=~/baspacho nox
-
-  py39_linux:
+      - <<: *run_lie_group_tests
+
+  py39_lie_tests:
     docker:
       - image: cimg/python:3.9
     steps:
       - checkout
       - <<: *install_suitesparse
       - <<: *install_functorch
       - <<: *build_baspacho_cpu
-      - <<: *install_nox
-      - run:
-          name: "Testing theseus"
-          command: |
-              export NOX_PYTHON_VERSIONS=3.9
-              pip install nox==2020.8.22
-              BASPACHO_ROOT_DIR=~/baspacho nox
+      - <<: *run_lie_group_tests
+
+  py310_lie_tests:
+    docker:
+      - image: cimg/python:3.10
+    steps:
+      - checkout
+      - <<: *install_suitesparse
+      - <<: *install_functorch
+      - <<: *build_baspacho_cpu
```

Comment:
You probably don't need baspacho to run lie group tests?

Reply:
Looks like we never made suitesparse optional, and pip install -e . fails.

Reply:
Ah okay yeah it would need extra things for the selective tests. We probably don't save that much time anyways by avoiding the install.